### ISSUE TITLE
Ability to remove dynamic partitions via Dagit

### ISSUE THREAD

**Issue by Daniel-Vetter-Coverwhale** [dagster-io/dagster#14026]
### What's the use case?

Sometimes you might need to remove a partition from a dynamic partition definition, in the case of an unexpected sensor catch or fat fingering in manually adding a partition and then you would need to remove it. Or maybe a new partition was added but you don't want it to be listed as the last partition when using LastPartitionMapping, and so you need to remove the actual last partition you want and then re-add it. 

### Ideas of implementation

Checkboxes next to the partitions in the asset catalogue page for partitioned assets and an action picker or entry in the Materialize dropdown that says something like "Remove Partitions", and exposing the action via graphql. 

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by tacastillo** [comment]
Thanks for raising this so we can track it!

I'm gonna link your original Slack thread here for posterity. https://dagster.slack.com/archives/C01U5LFUZJS/p1682946879458659

**Comment by brettallred** [comment]
Is there anyway to remove dynamic partitions in Dagster Cloud at all?

**Comment by sryza** [comment]
@brettallred alas there is not currently from the UI. You can write code that does this from a sensor, or using `DagsterInstance.delete_dynamic_partition`.

**Comment by dridk** [comment]
Could you show me an example ? 

How can I access to DagsterInstance from a job ?

**Comment by yuhan** [comment]
```python
@asset
def my_asset(context):
    context.instance.delete_dynamic_partition("my_partitions_def_name", "partition_key_to_delete")
```
apidoc: https://docs.dagster.io/_apidocs/internals#dagster.DagsterInstance.delete_dynamic_partition

similarly, you can access Dagster Instance in the context object in other entities like op, sensor, etc.

alternatively, you can also directly access dagster instance:
```python
from dagster import DagsterInstance
instance = DagsterInstance.get()
instance.delete_dynamic_partition("my_partitions_def_name", "partition_key_to_delete")
```

**Comment by sryza** [comment]
Here is a PR that will add the backend for this: https://github.com/dagster-io/dagster/pull/22185

@braunjj - thoughts on what this should look like in the frontend?

**Comment by braunjj** [comment]
How about a simple approach to this for now. 
- New Delete Partitions option in the actions dropdown
- Confirmation modal appears
- Defaults to the selected partition, but you can add or remove other partitions in the confirmation dialog 
- Delete button confirms the total number of partitions to be deleted
![image](https://github.com/dagster-io/dagster/assets/2798333/7e7c1352-8629-4df6-b289-f2f31f83cd77)
![image](https://github.com/dagster-io/dagster/assets/2798333/aaaade3c-1c70-4935-bf35-951e68b41e0c)

**Comment by sryza** [comment]
@braunjj something that strikes me as a little weird about this is that it's distant from the places where you can add dynamic partitions. If we go this route, would it make sense to also allow that within this dialog?

Something else to think about is that the dynamic partitions definition that partitions are being deleted for can be shared across multiple assets. So we should at least make it clear that it can have effects beyond the single asset being viewed here.

**Comment by andoni-garcia-fgp** [comment]
Is there any update here? It strikes me as very asymmetrical that we can add dynamic partitions but not remove them from the UI. We use dynamic string partitions and if a user enters a typo in their partition name, they cannot remove that erroneous partition. Further, running the asset over all partitions will also continue to include it.

**Comment by braunjj** [comment]
Hey @andoni-garcia-fgp, completely agree. The team started work to implement the UI this week so it should be shipping soon.

**Comment by Superjisan** [comment]
In the graphQL api, you can run this

```
mutation DeleteDynamicPartition(
  $partitionsDefName: String!, 
  $partitionKeys: [String!]!,
  $repositorySelector:RepositorySelector!
) {
  deleteDynamicPartitions(partitionsDefName: $partitionsDefName, partitionKeys: $partitionKeys, repositorySelector:$repositorySelector) {
    __typename
    ... on DeleteDynamicPartitionsSuccess {
      partitionsDefName
    }
    ... on PythonError {
      message
      stack
    }
  }
}
```

the query variables are 
```
{
  "partitionsDefName": "partition_def_name",
  "partitionKeys": ["bad_key"],
  "repositorySelector": {
    "repositoryName": "__repository__",
    "repositoryLocationName":"dagster_project"
  }
}
```

**Comment by onefloid** [comment]
Just for clarification — and to make things easier for anyone coming after me: this feature was implemented in PR [#23462](https://github.com/dagster-io/dagster/pull/23462) and is available starting in [Dagster 1.8.2](https://docs.dagster.io/about/changelog#182-core--0242-libraries) or higher.
